ipwb/replay.py

```diff
@@ -237,9 +239,12 @@ def resolveMemento(urir, datetime):
 
 @app.route('/memento/<regex("[0-9]{1,14}"):datetime>/<path:urir>')
 def showMemento(urir, datetime):
-    (newDatetime, linkHeader, uri) = resolveMemento(urir, datetime)
+    resolvedMemento = resolveMemento(urir, datetime)
```

Comment:
We can perhaps raise a custom `NoMementoFound` error from the `resolveMemento` function and catch it here. Having multiple different return signatures in a function is not a good practice.

Reply:
That sounds like a good idea to improve this. However, lets get this merged in to fix the big issue.

Reply:
Then add a TODO note in the `resolveMemento` function and fix the following line too before merging it:

https://github.com/oduwsdl/ipwb/blob/6b8ac9dd6b38128d7fa0db96f0b4a51bbde2a09a/ipwb/replay.py#L313

Reply:
@ibnesayeed If we use an Exception for this (what you meant by `NoMementoFound` "error"), I think there will be an issue answering the Flask route with the expected response object...unless we have the except return a Flask.response with status of 404 and other attributes about the request. No handling the exception in the...exception handler...sounds like it is not the best way to go.

Reply:
`NoMementoFound` (or something like that) will be a custom exception. The exception will be handled by calling `resolveMemento` in a try block. The `404` response can be generated in the corresponding except block. If there is some information that is only available in `resolveMemento` then it can be placed inside of the error message of the exception.

Reply:
`resolveMemento()` ideally returns `(newDatetime, linkHeader, uri)` tuple. If we throw an exception there as well as handle it within `resolveMemento()`, what do you anticipate `resolveMemento()` returning? Currently it creates a `flask.Response` and returns this. We type check in the caller (`showMemento()` and  `queryTimeGate()`) currently, which I know is a sub-optimal redundancy, but I don't think that the refactoring you suggest will resolve the issue of `resolveMemento()` returning two different types (a `flask.Response` or a `tuple`).

Reply:
The purpose of raising an exception to come out of the `resolveMemento` without any return value as the expected tuple cannot be populated there. We do not want to handle that exception in the same function from where it is being raised. We are already doing that inline handling, which is not good. The exception will be handled in the caller functions (i.e., `showMemento` and `queryTimeGate`).

Reply:
If we do not want to handle those exceptions, we might be able to raise and existing `NotFound` exception from the `werkzeug.exceptions`. I am not sure if Flask will handle it as it bubbles up. If it does on the application level then it should work without explicitly handling it every time `resolveMemento` is called.